enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,613 @@
+
+Functionality will be implemented by placing in `/etc/greenboot/green.d` and `/etc/greenboot/red.d`
+bash scripts containing with simple logic or, if needed,
+executing commands `microshift greenboot green` and `microshift greenboot red` in case of needing to put more information into file with action for next boot.
+Alternatively, a symlinks to `microshift` binary can be made and MicroShift modified to run specific command depending on content of argv[0] (just like BusyBox).
+
+File containing said information should not be part of MicroShift data directory
+as we don't to be a part of backup.
+
+### Backup and restore of MicroShift data
+
+Decision whether to backup or restore is based on file persisted during previous boot (see "Integration with greenboot").
+When file is read to make a decision (but not during dry run), it shall be removed.
+
+As a result of investigation and aiming for simplicity for initial implementation,
+it was decided that backing up MicroShift's data will be done by leveraging using copy-on-write (CoW) functionality.
+
+CoW is a feature of filesystem (supported by XFS and Btrfs) and it can be used by providing a `--reflink=` param to `cp` option.
+`--reflink=auto` will be used over `--reflink=always` to gracefully fall back to regular copying on filesystems
+not supporting CoW (ext4, ZFS).
+Backup will be done in `/var/lib/microshift.bak` or similar - it needs to be within the same filesystem/mount.
+Only one backup will be kept. Consequent boots will overwrite the backup dir (only if the action is "backup").
+
+End user documentation needs to include guidance on setting up filesystem to fullfil requirements for using copy-on-write (e.g. making sure some filesystem options are not disabled).
+
+### MicroShift version metadata persistence
+
+When MicroShift is up and running healthy, it will persist its own version and ID of current ostree commit into data dir as a file, e.g.:
+```json
+{
+  "microshift": "4.14.0",
+  "ostree_commit": "1a2b",
+}
+```
+
+MicroShift version will be used on consequent reboots to determine:
+- is upgrade is allowed or blocked
+- should storage migration be performed
+
+ID of ostree commit will be used to... TODO: Think if we can use it for "System rollback and failed data restore" scenario
+
+### Allowing and blocking upgrades (change of MicroShift version)
+
+Decision to allow or block an upgrade (change) will be based on persisted MicroShift version.
+- If binary's version is newer, allow storage migration.
+- If binary's version is older, refuse to start MicroShift.
+- If persisted version is on a list (embedded in binary) of blocked version migrations, refuse to start MicroShift.
+- If persisted version is on a list (embedded in binary) of allowed version migrations, allow storage migration.
+- If persisted version is missing and version of binary is 4.14, assume 4.13 was persisted and allow storage migration.
```

Comment:
In 4.15, would we make a different assumption? Or would we consider it an error to not have the version information persisted? Does it make sense to say that if there is no version information we will assume 4.13 and apply the rest of the rules (skews, etc.) based on that assumption?

Reply:
I addressed it in flowchart, but forgot to change it here.
Yes: missing version == 4.13, and then binary decides if the skew is too big or not